src/middleware/insights.ts

```diff
@@ -0,0 +1,121 @@
+import { Middleware } from '.';
+import { InsightsClient, InstantSearch } from '../types';
+import { getInsightsAnonymousUserToken } from '../helpers';
+
+export type InsightsProps = {
+  insightsClient: InsightsClient;
+};
+
+export type CreateInsightsMiddleware = (props: InsightsProps) => Middleware;
+
+export const createInsightsMiddleware: CreateInsightsMiddleware = props => {
+  const { insightsClient } = props;
+  if (!insightsClient) {
+    throw new Error(
+      'passing insightsClient to instantsearch is required for insightsMiddleware'
+    );
+  }
+  return ({ instantSearchInstance }) => {
```

Comment:
```suggestion
export type InsightsProps = {};

export type CreateInsightsMiddleware = (props?: InsightsProps) => Middleware;

export const createInsightsMiddleware: CreateInsightsMiddleware = (props = {}) => {
  return ({ instantSearchInstance }) => {
    const insightsClient = instantSearchInstance.insightsClient;
    if (!insightsClient) {
      throw new Error(
        'passing insightsClient to instantsearch is required for insightsMiddleware'
      );
    }
```

I expected more something like this (still needs a test with InstantSearch with insightsClient & InstantSearch without insightsClient)

I left the props for RFC-37

Reply:
Oh, before, I didn't read your message in the exception carefully.

As I commented in the RFC,

> My rationale behind this is to move insightsClient totally from instantsearch initialisation to the middleware (eventually).

I want to get rid of `insightsClient` from the `instantSearchInstance` in the end. What do you think?

Reply:
Ah like that! I think it does make sense, but then the error message should be changed, so that this insightsClient becomes required, and it overrides the one set by the constructor (which should give a warning if they don’t match)

Reply:
just updated the error message: https://github.com/algolia/instantsearch.js/pull/4446/commits/20254cf18539b16ab1bff92907c40537ffd852d8